File: javascript/tennis.test.js
import {Tennis} from "./tennis";

describe('score', () => {
    let tennis = new Tennis('joey', 'tom');
    beforeEach(() => {
        tennis = new Tennis('joey', 'tom');
    });

    function scoreShouldBe(expected) {
        expect(tennis.score()).toBe(expected);
    }

    it('should be love all', () => {
        scoreShouldBe('love all');
    });
    it('should be fifteen love', () => {
        givenFirstPlayerScoreTimes(1);
        scoreShouldBe('fifteen love');
    });

    function givenFirstPlayerScoreTimes(times) {
        for (let i = 0; i < times; i++) {
            tennis.firstPlayerScore();
        }
    }

    it('should be thirty love', () => {
        givenFirstPlayerScoreTimes(2);
        scoreShouldBe('thirty love');
    });

    it('should be forty love', () => {
        givenFirstPlayerScoreTimes(3);
        scoreShouldBe('forty love');
    });

    it('should be love fifteen', () => {
        givenSecondPlayerScoreTimes(1);
        scoreShouldBe('love fifteen');
    });

    function givenSecondPlayerScoreTimes(times) {
        for (let i = 0; i < times; i++) {
            tennis.secondPlayerScore();
        }
    }

    it('should be love thirty', () => {
        givenSecondPlayerScoreTimes(2);
        scoreShouldBe('love thirty');
    });

    it('should be fifteen all', () => {
        givenSecondPlayerScoreTimes(1);
        givenFirstPlayerScoreTimes(1);
        scoreShouldBe('fifteen all');
    });

    it('should be thirty all', () => {
        givenSecondPlayerScoreTimes(2);
        givenFirstPlayerScoreTimes(2);
        scoreShouldBe('thirty all');
    });

    function givenDeuce() {
        givenSecondPlayerScoreTimes(3);
        givenFirstPlayerScoreTimes(3);
    }

    it('should be deuce', () => {
        givenDeuce();
        scoreShouldBe('deuce');
    });

    it('should be first player adv', () => {
        givenDeuce();
        givenFirstPlayerScoreTimes(1);
        scoreShouldBe('joey adv');
    });

    it('should be second player adv', () => {
        givenDeuce();
        givenSecondPlayerScoreTimes(1);
        scoreShouldBe('tom adv');
    });

    it('should be second player win', () => {
        givenDeuce();
        givenSecondPlayerScoreTimes(2);
        scoreShouldBe('tom win');
    });


});

File: java/src/TennisTest.java
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class TennisTest {

    private final Tennis tennis = new Tennis("joey", "tom");

    @Test
    public void love_all() {
        scoreShouldBe("love all");
    }

    @Test
    public void fifteen_love() {
        givenFirstPlayerScoreTimes(1);
        scoreShouldBe("fifteen love");
    }

    @Test
    public void thirty_love() {
        givenFirstPlayerScoreTimes(2);
        scoreShouldBe("thirty love");
    }

    @Test
    public void forty_love() {
        givenFirstPlayerScoreTimes(3);
        scoreShouldBe("forty love");
    }

    @Test
    public void love_fifteen() {
        givenSecondPlayerScoreTimes(1);
        scoreShouldBe("love fifteen");
    }

    @Test
    public void love_thirty() {
        givenSecondPlayerScoreTimes(2);
        scoreShouldBe("love thirty");
    }

    @Test
    public void fifteen_all() {
        givenSecondPlayerScoreTimes(1);
        givenFirstPlayerScoreTimes(1);
        scoreShouldBe("fifteen all");
    }

    @Test
    public void thirty_all() {
        givenSecondPlayerScoreTimes(2);
        givenFirstPlayerScoreTimes(2);
        scoreShouldBe("thirty all");
    }

    @Test
    public void deuce() {
        givenDeuce();
        scoreShouldBe("deuce");
    }

    @Test
    public void first_player_adv() {
        givenDeuce();
        givenFirstPlayerScoreTimes(1);
        scoreShouldBe("joey adv");
    }

    @Test
    public void second_player_adv() {
        givenDeuce();
        givenSecondPlayerScoreTimes(1);
        scoreShouldBe("tom adv");
    }

    @Test
    public void second_player_win() {
        givenDeuce();
        givenSecondPlayerScoreTimes(2);
        scoreShouldBe("tom win");
    }

    private void givenDeuce() {
        givenSecondPlayerScoreTimes(3);
        givenFirstPlayerScoreTimes(3);
    }

    private void givenFirstPlayerScoreTimes(int times) {
        for (int i = 0; i < times; i++) {
            tennis.firstPlayerScore();
        }
    }

    private void givenSecondPlayerScoreTimes(int times) {
        for (int i = 0; i < times; i++) {
            tennis.secondPlayerScore();
        }
    }

    private void scoreShouldBe(String expected) {
        Assertions.assertEquals(expected, tennis.score());
    }
}
